Gait selection develop

Olavhaasie opened the pull request:
Closes PM-474

## Description
This is the *very big* PR that was coming for a long time now. Every change in this PR have been approved as separate steps, but it's nice to test everything together.

The build is unfortunately failing, because it uses the develop version of the downstream packages: monitor and simulation, which do not have the necessary changes.

## Changes
Some of the most important changes
* Removed `march_state_machine` and `march_gait_scheduler` package in favor of integrating those functionalities into the gait selection
* Gait state machine is generated on startup
* Gaits use graph descriptions in their configuration file to describe behavior
* No more intermediate action servers, instead 2 new topics
  - `CurrentGait` publishes details about every performed gait/subgait
  - `CurrentState` publishes the current state the state machine is in
* Gaits can now implement a `GaitInterface` to perform gaits instead of only preprogrammed trajectories

## Related PRs
* https://github.com/project-march/monitor/pull/68
* https://github.com/project-march/simulation/pull/63
* https://github.com/project-march/gait-files/pull/212

## Testing
Checkout the other PRs and start experimenting

Olavhaasie committed:
Add gait interface

Olavhaasie committed:
WIP: add state machine

Olavhaasie committed:
Add getting start and end transitions

Olavhaasie committed:
WIP: generate idle and gait transitions

Olavhaasie committed:
Add generating home gaits

Olavhaasie committed:
Add named positions

Olavhaasie committed:
Split up generate

Olavhaasie committed:
Add state machine input

Olavhaasie committed:
Implement idle state processing

Olavhaasie committed:
Add gait state processing

Olavhaasie committed:
Add get possible gaits service

Olavhaasie committed:
Removed starting position

Olavhaasie committed:
Make logging prettier

Olavhaasie committed:
Add forcing sm to unknown

Olavhaasie committed:
Implement update loop on subgaits

Olavhaasie committed:
Add home gait trajectory generation

Olavhaasie committed:
Fix suggestions

Olavhaasie committed:
Merge pull request #559 from project-march/new-gait-selection

New gait selection

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/gait-state-machine-flow

Olavhaasie committed:
Fix finishing of home gait

Olavhaasie committed:
Add update rate parameter

Olavhaasie committed:
Add try except for sleeping

Olavhaasie committed:
Fix transition to unknown

Olavhaasie committed:
Add trajectory scheduler

Olavhaasie committed:
Remove state machine and scheduler from launch

Olavhaasie committed:
Apply suggestions

Olavhaasie committed:
Add check for stoppable graph

Olavhaasie committed:
Add allow for stopping gait sooner

Olavhaasie committed:
Rename variable

Olavhaasie committed:
Rename default values

Olavhaasie committed:
Add start option to return trajectory

Olavhaasie committed:
Merge pull request #562 from project-march/feature/gait-state-machine-flow

gait-selection-develop: gait state machine flow

Olavhaasie committed:
Add error subscriber and responses

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Remove balance gait for the time being

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-507-error-handling

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-506-trajectory-scheduling

Olavhaasie committed:
Remove unused args

Olavhaasie committed:
Remove march_gait_scheduler package

Olavhaasie committed:
Add new message descriptions for current gait and state

Olavhaasie committed:
Add state publisher

Olavhaasie committed:
Add gait callbacks and generalize callbacks

Olavhaasie committed:
Add gait type and subgait name to setpoints gait

Olavhaasie committed:
Merge pull request #567 from project-march/feature/PM-507-error-handling

[gait-selection-develop] Feature: error handling

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-506-trajectory-scheduling

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-541-current-gait

Olavhaasie committed:
Fix current gait topic for gain scheduling

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-506-trajectory-scheduling

Olavhaasie committed:
Remove leftovers from action servers and clients

Olavhaasie committed:
Add msg header

Olavhaasie committed:
Add version to current gait

Olavhaasie committed:
Fix esp topics

Olavhaasie committed:
Add duration

Olavhaasie committed:
Merge pull request #568 from project-march/feature/PM-506-trajectory-scheduling

[gait-selection-develop] Feature: trajectory scheduling

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-541-current-gait

Olavhaasie committed:
Add subgait name to home gait

Olavhaasie committed:
Add sound loading in gait selection

Olavhaasie committed:
Remove march_state_machine package

Olavhaasie committed:
Add callbacks for playing sounds

Olavhaasie committed:
Add gait types for idle positions

RutgerVanBeek committed:
Merge pull request #571 from project-march/feature/PM-541-current-gait

[gait-selection-develop] Feature: current gait

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Rename is_idle

Olavhaasie committed:
Split up main into funtions

RutgerVanBeek committed:
Merge pull request #575 from project-march/feature/PM-508-playing-sounds

[gait-selection-develop] Feature: playing sounds

Olavhaasie committed:
Implement GaitInterface on balance

Olavhaasie committed:
Add balance gait to gait selection

Olavhaasie committed:
Fix gait selection tests

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-542-balance-gait

Olavhaasie committed:
Fix flake8

Olavhaasie committed:
Fix catkin lint

Olavhaasie committed:
Apply suggestions

Olavhaasie committed:
Merge pull request #577 from project-march/feature/PM-542-balance-gait

[gait-selection-develop] Feature: balance gait

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Add increase and decrease transitions

Olavhaasie committed:
Make only the to transition compulsory

Olavhaasie committed:
Exclude size transitions

Olavhaasie committed:
Make transition subgait accept subgaits instead of gait selection

Olavhaasie committed:
Implement subgait transition

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Merge branch 'gait-selection-develop' into feature/PM-543-subgait-transitions

Olavhaasie committed:
Add timestamp on header

Olavhaasie committed:
Fix transition steps by using correct subgaits

Olavhaasie committed:
Disable transitioning during a transition

Olavhaasie committed:
Remove else statement to reduce clutter

Olavhaasie committed:
Give stop priority over transition

Olavhaasie committed:
Fix crashing with stopping and transitions

Olavhaasie committed:
Merge pull request #581 from project-march/feature/PM-543-subgait-transitions

[gait-selection-develop] Feature: subgait transitions

RutgerVanBeek commented on march_gait_selection/src/march_gait_selection/dynamic_gaits/balance_gait.py:
```diff
@@ -191,22 +159,68 @@ def construct_subgait(self, capture_point_leg_name, subgait_name):
+        return self._default_walk.final_position
+
+    def start(self):
+        self._current_subgait = self._default_walk.graph.start_subgaits()[0]
```
Shouldn't you set the duration here as well.

Olavhaasie replied:
Oopsie this quite a big mistake and I have no idea how this actually works

RutgerVanBeek commented on march_gait_selection/src/march_gait_selection/state_machine/gait_interface.py:
```diff
@@ -0,0 +1,68 @@
+
+    @property
+    def subgait_name(self):
+        """Returns the name of the currently executing trajectory."""
+        return None
```
```suggestion
        """Returns the name of the currently executed trajectory."""
        return None
```

RutgerVanBeek commented on march_gait_selection/src/march_gait_selection/state_machine/gait_state_machine.py:
```diff
@@ -0,0 +1,260 @@
+
+    def _handle_input(self):
+        if self._input.stop_requested() or self._should_stop:
+            self._should_stop = False
```
Shouldn't you keep this True if stop is rejected?

Olavhaasie replied:
No otherwise it will still be True when the gait ends and you start a new gait. Furthermore, a gait rejects stop when it is not possible to stop the gait at all.

RutgerVanBeek commented on march_gait_selection/src/march_gait_selection/state_machine/state_machine_input.py:
```diff
@@ -0,0 +1,113 @@
+                                                               GaitInstructionResponse,
+                                                               queue_size=20)
+
+    def get_transition_request(self):
+        """Used to return the transition request as an enum.
+
+        :returns TransitionRequest
+        """
+        if self._transition_index == GaitInstruction.INCREMENT_STEP_SIZE:
+            return TransitionRequest.INCREASE_SIZE
+
+        if self._transition_index == GaitInstruction.DECREMENT_STEP_SIZE:
+            return TransitionRequest.DECREASE_SIZE
+
+        return TransitionRequest.NONE
```
Aren't you always returning the same integer here? Or is this for modularity?

Olavhaasie replied:
I don't really understand your question. It returns an enum, which depends on the transition index, when there is no transition it always returns `NONE`

RutgerVanBeek commented on march_gait_selection/src/march_gait_selection/state_machine/trajectory_scheduler.py:
```diff
@@ -0,0 +1,30 @@
+        :param JointTrajectory trajectory: a trajectory for all joints to follow
+        """
+        self._failed = False
+        goal = FollowJointTrajectoryGoal()
```
It is nice to also set the stamp

Olavhaasie replied:
The stamp in this message is used for planning the trajectory, when I add the current time here it will complain that the time of the first point is in the past. So it should be set to 0.

RutgerVanBeek reviewed (changes requested):
Mainly just questions

RutgerVanBeek commented:
![Screenshot from 2020-08-25 20-05-12](https://user-images.githubusercontent.com/51922475/91211130-8da23e00-e70e-11ea-80cb-49a54218c0ab.png)
This happend while doing tilted path left start. However I could not reproduce this. 

Furthermore there is a lot more info in the terminal, what do we think about that, is it too spammy? @Roelemans @JorisWeeda

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie committed:
Add timestamp to trajectory header

Olavhaasie committed:
Correctly init subgait duration

Olavhaasie committed:
Revert "Add timestamp to trajectory header"

This reverts commit 27ab1980f75532fc4ad6e1cc60f4d9f94310b46f.

Olavhaasie committed:
Fix sounds arg

JorisWeeda commented:
> Furthermore there is a lot more info in the terminal, what do we think about that, is it too spammy? @Roelemans @JorisWeeda

I don't really mind the extra information. Compared to the old state machine it only has some additional information at the start and end of the gait (requested, accepted, executed).

Roelemans reviewed (changes requested):
I have found two flaws:
1) You are able to change versions of a gait while that gait is being executed. This results in a midgait transition from, for example, "MV_walk_leftswing_v2" to "MIV_final" version of right swing, even though these subgait versions do not necessarily connect.
2) You are able to change versions such that gait transitions are no longer correct. This is possbile with both the bend_knee_left_tilted_path gait and the walk gait (when I created a new version of right open where the start did not match with standing, it allowed me to change it and execute it without problems)  

Ideally, to solve 1), you would allow changing only subgait versions of the gaits that are not currently being executed. However, you can also just disallow all version changes while not idle.

Olavhaasie committed:
Do not allow changing subgait versions while gait is active

Olavhaasie committed:
Add starting and final position to subgait

Olavhaasie committed:
Implement check for start and final position of subgait versions

Olavhaasie committed:
Merge branch 'develop' into gait-selection-develop

Olavhaasie commented:
> You are able to change versions of a gait while that gait is being executed. This results in a midgait transition from, for example, "MV_walk_leftswing_v2" to "MIV_final" version of right swing, even though these subgait versions do not necessarily connect.

Fixed this as you suggested by disallowing changing subgait versions of active gaits.

> You are able to change versions such that gait transitions are no longer correct. This is possbile with both the bend_knee_left_tilted_path gait and the walk gait (when I created a new version of right open where the start did not match with standing, it allowed me to change it and execute it without problems)

This is very good that you noticed this, since I forgot to implement checking changing subgait versions that are connected to `start` or `end`. I have now added two extra checks that make sure the new subgait version has the same positions, so that the state machine does not have to be re-generated. In the future you could of course build a way of notifying the state machine of such a change and re-generate the state machine.

Roelemans commented on march_shared_classes/src/march_shared_classes/gait/gait.py:
```diff
@@ -115,14 +119,30 @@ def set_subgait_versions(self, robot, gait_directory, version_map):
+
+                    if old_subgait.final_position != new_subgait.final_position:
+                        raise NonValidGaitContent(
+                            'The final position of new version {gait} {subgait} does not match'.format(
```
```suggestion
                            msg='The final position of new version {gait} {subgait} does not match'.format(
```

Roelemans commented on march_shared_classes/src/march_shared_classes/gait/gait.py:
```diff
@@ -115,14 +119,30 @@ def set_subgait_versions(self, robot, gait_directory, version_map):
+
+                    if old_subgait.starting_position != new_subgait.starting_position:
+                        raise NonValidGaitContent(
+                            'The starting position of new version {gait} {subgait} does not match'.format(
```
```suggestion
                            msg='The starting position of new version {gait} {subgait} does not match'.format(
```

Roelemans reviewed (changes requested):
Nice job! One last thing :innocent: 
![image](https://user-images.githubusercontent.com/47626418/92414034-d87e7580-f152-11ea-9a00-757b978a2398.png)

RutgerVanBeek commented on march_shared_classes/src/march_shared_classes/gait/subgait.py:
```diff
@@ -298,6 +298,16 @@ def get_joint_names(self):
+    @property
+    def starting_position(self):
+        """Returns a dictionary of joint positions at the start of this subgait."""
+        return dict([(joint.name, joint.setpoints[0].position) for joint in self.joints])
```
```suggestion
        return {joint.name: joint.setpoints[0].position for joint in self.joints}
```
It think this is the prettier way to do it.

RutgerVanBeek commented on march_shared_classes/src/march_shared_classes/gait/subgait.py:
```diff
@@ -298,6 +298,16 @@ def get_joint_names(self):
+    @property
+    def final_position(self):
+        """Returns a dictionary of joint positions at the end of this subgait."""
+        return dict([(joint.name, joint.setpoints[-1].position) for joint in self.joints])
```
```suggestion
        return {joint.name: joint.setpoints[-1].position for joint in self.joints}
```

RutgerVanBeek reviewed (changes requested):
Is it on purpose that the time is not automatically running, when starting? I would prefer it do so.

Olavhaasie commented:
> Is it on purpose that the time is not automatically running, when starting? I would prefer it do so.
Do you mean that the simulation is paused? Because that should not be the default

Olavhaasie committed:
Apply suggestions from code review

Co-authored-by: Roelemans <47626418+Roelemans@users.noreply.github.com>
Co-authored-by: RutgerVanBeek <51922475+RutgerVanBeek@users.noreply.github.com>

RutgerVanBeek commented:
> > Is it on purpose that the time is not automatically running, when starting? I would prefer it do so.
> 
> Do you mean that the simulation is paused? Because that should not be the default

Last time I also wanted to test balance and checked out Roel's turn-off-com-branch, because of that didn't default to unpause. So it actually was allright when I check out the right branch in simulation now.
